### ISSUE TITLE
Reproducibility Issues with DeepLabV3Plus, PSPNet

### ISSUE THREAD

**Issue by 42nick** [qubvel/segmentation_models.pytorch#465]
Hello, 

i have problems with getting reproducible results with DeepLabV3Plus and PSPNet when reinitializing them. I followed [PyTorch Page](https://pytorch.org/docs/stable/notes/randomness.html) to insure the reproducibility. 

I wrote a minimal example to reproduce this issue.
For Unet or Unetplusplus (comment out the ) I can restart the script, create two model instances m1 and m2 and train each of them. The loss and the predictions with or with out gradient calculations is the same. 

When I start the script with the DeepLabV3Plus uncommented the losses are different for the two model instances, even if the predictions at the beginning where the same. 

Does somebody know why different DeepLabV3Plus-models behave differently after one training epoch while the Unet-models behave exactly the same? 

Tested with:

- python3.6
- torch 1.7.0+cu110 and torch 1.4.0+cu100

```python
import os
import random

import numpy as np
import segmentation_models_pytorch as smp
import torch
from torch import nn
from torch.optim import SGD, Optimizer


def set_seed(seed=42):
    random.seed(seed)
    os.environ["PYTHONHASHSEED"] = str(seed)
    np.random.seed(seed)
    torch.manual_seed(seed)
    torch.cuda.manual_seed(seed)
    torch.cuda.manual_seed_all(seed)
    torch.backends.cudnn.benchmark = False
    torch.backends.cudnn.deterministic = True


class Trainer:
    def __init__(self, model, opt, lossfn) -> None:
        self.model: nn.Module = model
        self.optim: Optimizer = opt
        self.lossfn = lossfn

    def trainEpoch(self, x, y):
        self.model.train()
        self.optim.zero_grad()
        prediction = self.model.forward(x)
        loss = self.lossfn(prediction, y)
        loss.backward()
        self.optim.step()
        return loss, prediction


### reproduceable architectures
# arch = smp.Unet
# arch = smp.UnetPlusPlus
# arch = smp.PAN

### not reproduceable architectures
# arch = smp.PSPNet
arch = smp.DeepLabV3Plus

args = dict(encoder_name="resnet18", encoder_weights="imagenet", classes=2)
print(arch)

# creating architectures with setting the seed in advance
set_seed()
m1 = arch(**args)
set_seed()
m2 = arch(**args)

# generating example input
img = torch.rand(2, 3, 128, 128)
gt = (torch.rand(2, 128, 128) < 0.5).long()  # ground truth

# get one prediction with no gradient calculation
m1.eval()
m2.eval()
with torch.no_grad():
    # # with torch.cuda.amp.autocast():
    p1 = m1(img)
    p2 = m2(img)

# train one epoch with each model
opt = SGD(params=m1.parameters(), lr=0.001)
trainer1 = Trainer(m1, opt, lossfn=nn.CrossEntropyLoss())
loss1, prediction1 = trainer1.trainEpoch(img, gt)

opt2 = SGD(params=m2.parameters(), lr=0.001)
trainer2 = Trainer(m2, opt2, lossfn=nn.CrossEntropyLoss())
loss2, prediction2 = trainer2.trainEpoch(img, gt)

# compare results
print("Prediction at the beginning ist the same: ", (p1 == p2).all())
print("Traingsloss is the same: ", loss1 == loss2)
print("Difference between predictions in training process: ", (prediction1 - prediction2).sum())
# print(loss1)
```

**Comment by lliuz** [comment]
The reason is unets use nearest interpolation which is deterministic, while others use bilinear interpolation which is non-deterministic.
You can refer to https://discuss.pytorch.org/t/non-deterministic-behavior-of-pytorch-upsample-interpolate/42842.

**Comment by 42nick** [comment]
Thanks a lot.
